### PR TITLE
add a QueryManyUniqueIter

### PR DESCRIPTION
**Pull request by Victoronz**
# Objective

~Requires #13476~.

Currently we cannot iterate mutably over `QueryManyIter` even if all entities we pass are unique.

## Solution

Add a `QueryManyUniqueIter` wrapper around `QueryManyIter` for full `Iterator` trait iteration.
It can be constructed via either the `dedup_entities` or `entities_all_unique` method on `QueryManyIter`.
`dedup_entities` deduplicates the internal `entity_iter`, retaining order.
`entities_all_unique` takes `self`, checks for uniqueness, returns `QueryManyUniqueIter` on success, and `QueryManyIter` (`Self`) on failure.
Since `QueryManyUniqueIter` can iterate over mutable `QueryData` , it does not expose a `fetch_next` method.

## Testing

The example/doc test shows the method in use, also making sure `expect` can be called on the return `Result`.

## Changelog

Added `dedup_entities` and `entities_all_unique` methods on `QueryManyIter`.
Added `QueryManyUniqueIter` struct.

### CODE REVIEW

File: crates/bevy_ecs/src/query/iter.rs [hunk 1]
```diff
@@ -1161,6 +1239,118 @@ where
...
+    query_state: &'s QueryState<D, F>,
+}
+
+impl<'w, 's, D: QueryData, F: QueryFilter, I: Iterator> QueryManyUniqueIter<'w, 's, D, F, I>
+where
+    I::Item: Borrow<Entity>,
+{
+    // Entities are guaranteed to be unique, so no lifetime shrinking needed for mutable iteration.
+    #[inline(always)]
+    fn fetch_next(&mut self) -> Option<D::Item<'w>> {
```
[the remark this answers is not in the document]
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:1269]: IIRC, I was not sure whether the uniqueness property could simplify the `next()` call, so I kept them separate, but looking at it again, you're right, this can just be wrapper type! Good idea.

File: crates/bevy_ecs/src/query/iter.rs [hunk 2]
```diff
@@ -1123,6 +1124,83 @@ where
...
+    ///         **part_value += 1;
+    ///     }
+    /// }
+    /// #
+    /// # let mut schedule = Schedule::default();
+    /// # schedule.add_systems((system));
+    /// # schedule.run(&mut world);
+    /// ```
+    #[inline(always)]
+    pub fn entities_all_unique(
```
[the remark this answers is not in the document]
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:1173]: I agree with the methods to construct from unique iterators. What other iterator types would you suggest?

Deduplication can already be done on the iterator itself with `itertools`, so isn't as important to have as the others. Still a nice to have though.
I wonder whether a sort + iterative dedup would be faster than your `HashSet` version.

I think I'll make these separate follow-up PRs, since I have to keep these changes in line with #13443.

File: crates/bevy_ecs/src/query/iter.rs [hunk 3]
```diff
@@ -1076,11 +1077,17 @@ where
                 continue;
             }
 
-            let archetype = self
-                .archetypes
-                .get(location.archetype_id)
-                .debug_checked_unwrap();
-            let table = self.tables.get(location.table_id).debug_checked_unwrap();
+            let (archetype, table);
```
Comment (chescock) [inline crates/bevy_ecs/src/query/iter.rs:1080]: This is good cleanup, but could probably be a separate PR.

File: crates/bevy_ecs/src/query/iter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:1179]: I have added `dedup_entities` on `QueryManyIter` to this PR anyway.

**Review by chescock — APPROVED** [review]

**Comment by Victoronz** [comment]
I wonder whether it might be easier to have a direct `Query::iter_many_unique` method instead of these roundabout conversions on `QueryManyIter`.

File: crates/bevy_ecs/src/query/iter.rs (hunk 3 above)
Comment (Victoronz) [reply crates/bevy_ecs/src/query/iter.rs:1080]: Fair, I'll put it in a different PR.

**Comment by Victoronz** [comment]
I have come up with a more powerful, general solution, which I think I will put in a different PR. I'll keep this open as the simpler alternative.

This PR would also need to be updated with a `TrustedEq: Eq` trait, because even though we only use it for `Entity`, unsafe code must not rely on an `Eq` trait impl for correctness, because `Eq` itself does not contain a logical correctness guarantee.
That being said, this would be a simple unsafe marker trait impl on `Entity` stating the safety contract.

**Comment by Victoronz** [comment]
This PR is now superseded by the concept of #16547, though the actual conversion API would be a follow-up to that PR.

**Comment by Victoronz** [comment]
#16547 has merged.
